File: StatePattern/cpp/src/taxi.cpp
#include <iostream>
#include "header/taxi.h"

shared_ptr<State> OffDutyLights::hailed(Taxi *taxi){
    cout << "\tPassenger board on the taxi." << endl;
    return taxi->onDutyLight;
}

shared_ptr<State> OffDutyLights::drive(Taxi *taxi){
    cout << "\tNo passenger..." << endl;
    return taxi->offDutyLights;
}

shared_ptr<State> OffDutyLights::checkout(Taxi *taxi){
    cout << "\tNo passenger..." << endl;
    return taxi->offDutyLights;
}

shared_ptr<State> OnDutyLight::hailed(Taxi *taxi){
    cout << "\tAlready hailed." << endl;
    return taxi->onDutyLight;
}

shared_ptr<State> OnDutyLight::drive(Taxi *taxi){
    cout << "\tTake passenger to their destination." << endl;
    return taxi->arrived;
}

shared_ptr<State> OnDutyLight::checkout(Taxi *taxi){
    cout << "\tThe passenger's destination is not arrived." << endl;
    return taxi->onDutyLight;

}

shared_ptr<State> Arrived::hailed(Taxi *taxi){
    cout << "\tAlready hailed." << endl;
    return taxi->arrived;
}

shared_ptr<State> Arrived::drive(Taxi *taxi){
    cout << "\tWe are at your destination." << endl;
    return taxi->arrived;
}

shared_ptr<State> Arrived::checkout(Taxi *taxi){
    cout << "\tCheckout the bill." << endl;
    return taxi->offDutyLights;
}

Taxi::Taxi(){
    offDutyLights = make_shared<OffDutyLights>();
    onDutyLight = make_shared<OnDutyLight>();
    arrived = make_shared<Arrived>();
    state = offDutyLights;
}

shared_ptr<State> Taxi::getState(){
    return state;
}

void Taxi::hailed(){
    cout << "Someone hails the taxi." << endl;
    shared_ptr<State> previouState = state;
    state = state->hailed(this);
    if(previouState != state){
        state = state->drive(this);
    }
}

void Taxi::checkout(){
    cout << "The passenger checkout the bill." << endl;
    state = state->checkout(this);
}

File: StatePattern/python/Taxi.py
import abc

class State(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def hailed(self, taxi:'Taxi') -> 'State':
        return NotImplemented
    
    @abc.abstractmethod
    def drive(self, taxi:'Taxi') -> 'State':
        return NotImplemented
    
    @abc.abstractmethod
    def checkout(self, taxi:'Taxi') -> 'State':
        return NotImplemented
    

class OffDutyLights(State):
    def hailed(self, taxi: 'Taxi') -> State:
        print("\tPassenger board on the taxi.")
        return taxi._onDutyLight
    
    def drive(self, taxi: 'Taxi') -> State:
        print("\tNo passenger...")
        return self
    
    def checkout(self, taxi: 'Taxi') -> State:
        print("\tNo passenger...")
        return self
    

class OnDutyLight(State):
    def hailed(self, taxi: 'Taxi') -> State:
        print("\tAlready hailed.")
        return self
    
    def drive(self, taxi: 'Taxi') -> State:
        print("\tTake passenger to their destination.")
        return taxi._arrived
    
    def checkout(self, taxi: 'Taxi') -> State:
        print("\tThe passenger's destination is not arrived.")
        return self


class Arrived(State):
    def hailed(self, taxi: 'Taxi') -> State:
        print("\tAlready hailed.")
        return self
    
    def drive(self, taxi: 'Taxi') -> State:
        print("\tWe are at your destination.")
        return self
    
    def checkout(self, taxi: 'Taxi') -> State:
        print("\tCheckout the bill.")
        return taxi._offDutyLights
    

class Taxi:
    def __init__(self) -> None:
        self._offDutyLights = OffDutyLights()
        self._onDutyLight = OnDutyLight()
        self._arrived = Arrived()
        self._state = self._offDutyLights

    @property
    def state(self):
        return self._state
    
    def hailed(self):
        print("Someone hails the taxi.")
        previouState = self.state
        self._state = self._state.hailed(self)
        if previouState != self._state:
            self._state = self._state.drive(self)

    def checkout(self):
        print("The passenger checkout the bill.")
        self._state = self._state.checkout(self)